core/src/test/java/org/ldaptive/PooledConnectionFactoryTest.java

```diff
@@ -252,4 +254,82 @@ public void validatePeriodicallyWithTimeout()
       server.stop();
     }
   }
+
+
+  /**
+   * @throws  Exception  On test failure.
+   */
+  @Test(groups = "netty")
+  public void validateOnCheckout()
+    throws Exception
+  {
+    final SimpleNettyServer server = new SimpleNettyServer();
+    try {
+      final InetSocketAddress address = server.start();
+      final AtomicInteger validateCount = new AtomicInteger();
+      final PooledConnectionFactory factory = PooledConnectionFactory.builder()
+        .config(ConnectionConfig.builder()
+          .url(new LdapURL(address.getHostName(), address.getPort()).getHostnameWithSchemeAndPort())
+          .build())
+        .min(0)
+        .validator(new ConnectionValidator() {
+          @Override
+          public void applyAsync(final Connection conn, final Consumer<Boolean> function) {}
+
+          @Override
+          public Supplier<Boolean> applyAsync(final Connection conn)
+          {
+            return () -> false;
+          }
+
+          @Override
+          public Duration getValidatePeriod()
+          {
+            return Duration.ofMinutes(5);
+          }
+
+          @Override
+          public Duration getValidateTimeout()
+          {
+            return Duration.ofSeconds(5);
+          }
+
+          @Override
+          public Boolean apply(final Connection connection)
+          {
+            return validateCount.getAndIncrement() != 1;
+          }
+        })
+        .validateOnCheckOut(true)
```

Comment:
My only real concern here is around configuration and arriving at perhaps unexpected behavior in either of the following cases:

1. Setting `validateOnCheckOut(false)` and configuring a `ValidationExceptionHandler`.
2. Setting `validateOnCheckOut(true) and _not_ configuring a `ValidationExceptionHandler`.

The first configuration doesn't do anything, and the second causes an exception to be raised by the pool if the connection is bad. I _think_ the latter behavior is reasonable, but I think a friendlier behavior would be to simply keep trying, which you only get if you configure as you have here in this test case. Thus my biggest concern is the decoupling of these two flags and possibly adding configuration complexity. You might consider simply configuring the retry handler when setting the `validateOnCheckout` flag to true under the assumption that's what most people want; if they have some bizarre requirement to further customize the setter on the factory is still there to do as needed.

Reply:
I think for (1) the code should emit a warning, I'll make that change. For (2), you're correct, that particular configuration would allow a client to implement their own handling of `ValidationException`.  (Which I'm proposing be the default and you could argue is even currently the default now.)

I'm obviously avoiding any sort of _default_ configuration here because of the potential dangers with retry. The first being, what sort of default configuration would be reasonable? (retry once?) And the second being, retry has the real potential to delay feedback to the caller. In cases where you're already having trouble with connections, asking again (and again) can be much worse than simply returning control with an exception. Particularly when you're dealing with a  server that is slow to connect and/or slow to respond.

Reply:
I pushed commits to resolve this. I came to terms with using a default retry implementation.

Reply:
Seems good.